### ISSUE TITLE
Turkish İ/I corruption in article_parser.py (casefold) — and why a downstream hard-coded fix is a no-op

### ISSUE THREAD

**Issue by surmeliugur** [lonewolf-rd/CorpusCollector#6]
`src/helpers/article_parser.py:49` uses `full_text.casefold()`. For Turkish text this corrupts both dotted and dotless I, *systematically*:

```python
>>> "İSTANBUL IŞIK ÜNİVERSİTESİ".casefold()
'i̇stanbul işik üni̇versi̇tesi̇'   # 'İ' → 'i' + U+0307 (combining dot), 'I' → 'i' (should be 'ı')
```

Two separate problems:

**1. The corruption itself.** Every `İ` becomes `i` + U+0307 (a combining-dot residue that survives as a separate codepoint), and every `I` becomes `i` instead of `ı`. Since DergiPark academic titles/proper nouns are İ/I-heavy, the corruption is not random noise — it is a systematic shift affecting the same word classes, so a neural tokenizer can learn `i̇` sequences as spurious token material.

**2. The downstream fix ordering trap (the subtle part).** A hard-coded `replace("İ","i").replace("I","ı")` applied *after* `casefold()` is a **no-op**: at that point the text no longer contains `İ` or `I` at all. Measured:

```python
s = "IŞIK"
s.casefold().replace("İ","i").replace("I","ı")   # → 'işik'  (still wrong; U+0307 residues remain elsewhere)
s.replace("İ","i").replace("I","ı").lower()      # → 'ışık'  (correct)
```

So if the cleanup runs after the parser's `casefold`, the final dataset still carries the corruption even though a fix exists in the pipeline.

**Suggested actions:**

- In the parser, replace `casefold()`/`lower()` with Turkish-aware lowercasing: `text.replace("İ","i").replace("I","ı").lower()` (or `str.translate` for speed).
- **Retrospective diagnosis without re-crawling:** count U+0307 (`"̇"`) occurrences in the released corpus — every hit is a corrupted `İ`; the dotless-I damage can be estimated from tokens matching `işik`-class patterns. This measures the real extent instead of assuming "<1%".
- A note in the README for anyone building Turkish corpora: `casefold`/`lower` are unsafe for Turkish; this bites virtually every Turkish NLP pipeline at some point.

Happy to help verify the fix. (Found while cross-reading Morpheus for a tokenizer comparison; the same trap has bitten our own pipeline before.)

<details>
<summary>Türkçe özet</summary>

`src/helpers/article_parser.py:49` `full_text.casefold()` kullanıyor. Türkçede bu, noktalı ve noktasız I'yı *sistematik* bozar: `'İ'` → `'i' + U+0307` (birleşen nokta artığı ayrı kod noktası olarak kalır), `'I'` → `'i'` (doğrusu `'ı'`).

İki ayrı sorun:

1. **Bozulmanın kendisi:** DergiPark başlıkları/özel adlar İ/I-yoğun olduğundan bozulma rastgele gürültü değil — aynı kelime sınıflarını aynı yöne kaydıran sistematik bir kayma; neural tokenizer `i̇` dizilimlerini sahte token malzemesi olarak öğrenebilir.

2. **Düzeltme-sırası tuzağı (ince nokta):** `casefold` SONRASI uygulanan sabit-kodlu `replace("İ","i").replace("I","ı")` **no-op'tur** — o noktada metinde artık `İ`/`I` karakteri yoktur. Ölçüldü: `"IŞIK".casefold()` + düzeltme → `işik` (yanlış kalır); doğru sıra ÖNCE replace SONRA `lower()` → `ışık`.

**Öneriler:** parser'da Türkçe-duyarlı küçültme (önce replace, sonra lower); yayınlanmış korpusta **U+0307 sayımı** ile retrospektif teşhis (yeniden taramadan gerçek boyut ölçülür — "&lt;%1" varsayımı yerine); README'ye Türkçe korpus kuranlar için `casefold`/`lower` uyarısı.

</details>

**Comment by dfavenfre** [comment]
Thanks @surmeliugur, reproduced on CPython 3.12, the mechanics are exactly as you describe. Some scope, though, since the impact model matters more than the call site here.

This corpus is tokenizer training data, not model training data: 27 MB of DergiPark text against a ~10 GB total, so roughly 0.3% of the vocabulary-induction input. It also isn't the İ/I-dense 0.3%,  the extraction window runs `giriş` → `kaynaklar`, so titles, authors, affiliations and the reference list, i.e. exactly the proper-noun-heavy regions, never enter the file.

On downstream propagation: everything Morpheus ingests passes through `turkish_lower` in [TurkishMorpheus](https://github.com/lonewolf-rd/TurkishMorpheus) `/src/common/text_utils.py` before vocabulary induction. To be precise about what that does and doesn't buy, it's replace-then-lower and idempotent, so it prevents any *further* casing damage, but it does not retroactively repair an already-folded `I`. What it does mean is that the parser's fold isn't compounded anywhere downstream.

The losslessness property is also narrower than "the data is clean": Morpheus learns segmentation with a lossless surface↔segment conversion ([arXiv:2606.18717](https://arxiv.org/abs/2606.18717)), so it doesn't introduce canonical drift of its own and any residue stays a recoverable string-level artifact rather than getting baked into a lossy vocabulary. Practically, a `i̇`-bearing subword is a low-frequency variant sitting next to its clean form — closer to the typo-variant material a tokenizer benefits from than to a systematic vocabulary shift. Not an argument for keeping it, just for why this isn't corpus-invalidating.

The relevant question is whether the residue reached the vocabulary, and that's checkable rather than arguable. Live demo: https://huggingface.co/spaces/lonewolflab/morpheus-tr-demo — the `ışık` family, i.e. the exact `işik`-class you flagged, segments as:

<img width="2520" height="1027" alt="Image" src="https://github.com/user-attachments/assets/68a178f0-aaf1-4ea9-9ac8-ba1475ed62f0" />

Clean dotless-ı root as a single vocabulary item, stable across four inflections, `decode(encode(w)) == w` 4/4. A root that had absorbed `işik`/`i̇` material would surface as fragmentation or a competing variant token; it doesn't. This doesn't prove the 27 MB file itself is clean — at 0.3% it can't outvote the rest of the corpus either way, but it does show the failure mode the issue is actually about (spurious `i̇` token material, shifted word classes) did not materialize in the induced vocabulary.

Worth noting the roundtrip line separately: Morpheus is surface-preserving by construction ([arXiv:2606.18717](https://arxiv.org/abs/2606.18717)), so no canonical normalization is applied at encode time and any residue stays a recoverable string-level artifact instead of being baked into a lossy vocabulary, unlike the rule-based baseline, whose normalization can corrupt decoded text.

Fix as discussed: match on a folded probe, slice the original by span, so stored text keeps its casing. Will post the measured U+0307 count with the patch and add the `casefold`/`lower` note to the README. Closing.
